banzai/bpm.py

```diff
@@ -8,7 +20,72 @@ def group_by_keywords(self):
         return None
 
     def do_stage(self, images):
+        images_to_remove = []
         for image in images:
-            bpm_slices = array_utils.array_indices_to_slices(image.bpm)
-            image.bpm[image.data[bpm_slices] >= float(image.header['SATURATE'])] = 2
+            try:
+                self.add_bpm_to_image(image)
+            except MissingBPMError:
+                if self.pipeline_context.no_bpm:
+                    logger.info('BPM misisng but not required, falling back to empty BPM', image=image)
+                    pass
+                else:
+                    logger.error("Can't add BPM to image, stopping reduction", image=image)
+                    images_to_remove.append(image)
+                    continue
+            flag_bad_pixels(image)
+        for image in images_to_remove:
+            images.remove(image)
         return images
+
+    def add_bpm_to_image(self, image):
+        # Get the BPM filename
+        bpm_filename = self.get_bpm_filename(image)
+        # Check if file is missing
+        if bpm_filename is None:
+            logger.warning('Unable to find BPM in database', image=image)
+            raise MissingBPMError
+        # Load the BPM
+        bpm = load_bpm(bpm_filename)
+        # Check if the BPM is the right size
+        if not bpm_has_valid_size(bpm, image):
+            logger.warning('BPM shape mismatch', image=image)
+            raise MissingBPMError
+        # Add BPM to image and header info
+        image.bpm = bpm
+        image.header['L1IDMASK'] = (os.path.basename(bpm_filename), 'Id. of mask file used')
+
+    def get_bpm_filename(self, image):
+        return dbs.get_bpm_filename(image.telescope.id, image.ccdsum,
```

Comment:
get_bpm_filename could return an empty string instead of None perhaps?

Reply:
I've left this for now, did you have in mind that `add_bpm_to_image` would not short-circuit if the filename is None, but rather continue and try to load from an empty string? I couldn't get the logic for that to work in a nice way, but we can talk about it.